src/bezpath.rs

```diff
@@ -205,51 +212,39 @@ impl Mul<BezPath> for Affine {
 }
 
 struct BezPathSegs<'a> {
-    c: &'a [PathEl],
-    ix: usize,
-    start: Option<Vec2>,
-    last: Option<Vec2>,
+    c: std::slice::Iter<'a, PathEl>,
+    start: Vec2,
+    last: Vec2,
 }
 
 impl<'a> Iterator for BezPathSegs<'a> {
-    type Item = (usize, PathSeg);
-
-    fn next(&mut self) -> Option<(usize, PathSeg)> {
-        while self.ix < self.c.len() {
-            let ix = self.ix;
-            let el = self.c[ix];
-            self.ix += 1;
-            match el {
+    type Item = PathSeg;
+
+    fn next(&mut self) -> Option<PathSeg> {
+        for el in &mut self.c {
+            let (ret, last) = match *el {
                 PathEl::Moveto(p) => {
-                    self.start = Some(p);
-                    self.last = Some(p);
-                }
-                PathEl::Lineto(p) => {
-                    let seg = PathSeg::Line(Line::new(self.last.unwrap(), p));
-                    self.last = Some(p);
-                    return Some((ix, seg));
-                }
-                PathEl::Quadto(p1, p2) => {
-                    let seg = PathSeg::Quad(QuadBez::new(self.last.unwrap(), p1, p2));
-                    self.last = Some(p2);
-                    return Some((ix, seg));
+                    self.start = p;
+                    self.last = p;
+                    continue;
                 }
+                PathEl::Lineto(p) => (PathSeg::Line(Line::new(self.last, p)), p),
+                PathEl::Quadto(p1, p2) => (PathSeg::Quad(QuadBez::new(self.last, p1, p2)), p2),
                 PathEl::Curveto(p1, p2, p3) => {
-                    let seg = PathSeg::Cubic(CubicBez::new(self.last.unwrap(), p1, p2, p3));
-                    self.last = Some(p3);
-                    return Some((ix, seg));
+                    (PathSeg::Cubic(CubicBez::new(self.last, p1, p2, p3)), p3)
                 }
                 PathEl::Closepath => {
-                    let last = self.last.take();
-                    let start = self.start.take();
-                    if last != start {
-                        let seg = PathSeg::Line(Line::new(last.unwrap(), start.unwrap()));
-                        return Some((ix, seg));
+                    if self.last != self.start {
+                        return Some(PathSeg::Line(Line::new(self.last, self.start)));
                     }
+                    break;
```

Comment:
I think this is wrong. In the case of two successive triangles of the form "moveto(p1) lineto(p2) lineto(p3) moveto(p1) closepath", I think this makes it return `None` after the first triangle.

Of course, to really settle this question there should be a test :) The lack of tests is just laziness, as I was pounding out the code.

Reply:
Looks like you are right ... Well, I wrote a really dirty test to check for regressions but it looks like this is something I missed ... Will fix!